builder/anka/communicator.go

```diff
@@ -176,94 +160,112 @@ func (c *Communicator) UploadDir(dst string, src string, exclude []string) error
 			return dst.Chmod(si.Mode())
 		}
 
-		// Copy the entire directory tree to the temporary directory
-		if err := filepath.Walk(src, walkFn); err != nil {
+		err = filepath.Walk(src, walkFn)
+		if err != nil {
 			return err
 		}
 
-		// Determine the destination directory
-		// containerSrc := filepath.Join(c.ContainerDir, filepath.Base(td))
 		containerDst := dst
 		if src[len(src)-1] != '/' {
 			containerDst = filepath.Join(dst, filepath.Base(src))
 		}
 
 		log.Printf("from %#v to %#v", td, containerDst)
 
-		// Make the directory, then copy into it
 		command := fmt.Sprintf("set -e; mkdir -p %s; command cp -R %s/* %s",
 			containerDst, filepath.Base(td), containerDst,
 		)
-		err, _ = c.Client.Run(client.RunParams{
+
+		_, err = c.Client.Run(client.RunParams{
 			VMName:  c.VMName,
 			Command: []string{"bash", "-c", command},
 			Volume:  c.HostDir,
 		})
+
 		return err
-	} else {
-		return c.Client.Copy(client.CopyParams{
-			Src: src,
-			Dst: c.VMName + ":" + dst,
-		})
 	}
+
+	return c.Client.Copy(client.CopyParams{
+		Src: src,
+		Dst: c.VMName + ":" + dst,
+	})
 }
 
+// Download copies the file from the source to the destination
 func (c *Communicator) Download(src string, dst io.Writer) error {
 	log.Printf("Downloading file from VM: %s", src)
 
-	// Create a temporary file to store the download
+	c.configureAnkaCP()
+
 	tempfile, err := ioutil.TempFile(c.HostDir, "download")
 	if err != nil {
 		return err
 	}
+
 	defer os.Remove(tempfile.Name())
 	defer tempfile.Close()
 
 	if c.Config.UseAnkaCP {
-		err = c.Client.Copy(client.CopyParams{
+		err := c.Client.Copy(client.CopyParams{
 			Src: c.VMName + ":" + src,
 			Dst: tempfile.Name(),
 		})
+		if err != nil {
+			return err
+		}
 	}
 
 	log.Printf("Copying from %s to writer", tempfile.Name())
+
 	w, err := io.Copy(dst, tempfile)
 	if err != nil {
 		return err
 	}
 
-	if c.Config.UseAnkaCP == false {
-		errFindingFUSE := c.findFUSE()
-		if errFindingFUSE != nil {
-			c.Config.UseAnkaCP = true
-		}
-	}
-
-	if c.Config.UseAnkaCP == false {
-		err, _ = c.Client.Run(client.RunParams{
+	if !c.Config.UseAnkaCP {
+		_, err := c.Client.Run(client.RunParams{
 			VMName:  c.VMName,
 			Command: []string{"cp", src, "./" + path.Base(tempfile.Name())},
 			Volume:  c.HostDir,
 		})
+		if err != nil {
+			return err
+		}
 	}
 
 	log.Printf("Copied %d bytes", w)
+
 	return nil
 }
 
+// DownloadDir copies the source directory to the destination
 func (c *Communicator) DownloadDir(src string, dst string, exclude []string) error {
-	if c.Config.UseAnkaCP == false {
+	c.configureAnkaCP()
+
+	if !c.Config.UseAnkaCP {
+		return errors.New("communicator.DownloadDir isn't implemented")
+	}
+
+	return c.Client.Copy(client.CopyParams{
+		Src: c.VMName + ":" + src,
+		Dst: dst,
+	})
+}
+
+func (c *Communicator) configureAnkaCP() {
+	if !c.Config.UseAnkaCP {
```

Comment:
Is there a need to call this code more than once?

Reply:
I do think a singleton would be useful here. I can work on that

Reply:
Merged in new logic to solve this